substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/SubstrateNodeLLVMBuilder.java

```diff
@@ -59,6 +85,102 @@ public void emitCGlobalDataLoadAddress(CGlobalDataLoadAddressNode node) {
     @Override
     public Variable emitReadReturnAddress() {
         LLVMValueRef returnAddress = getLIRGeneratorTool().getBuilder().buildReturnAddress(getLIRGeneratorTool().getBuilder().constantInt(0));
-        return new LLVMUtils.LLVMVariable(returnAddress);
+        return new LLVMVariable(returnAddress);
+    }
+
+    private SubstrateLLVMGenerator getGenerator() {
+        return (SubstrateLLVMGenerator) gen;
+    }
+
+    @Override
+    protected int getParamIndex(int index) {
+        int offset = (getGenerator().isEntryPoint() ? 0 : LLVMFeature.SPECIAL_REGISTER_COUNT);
+        return offset + index;
+    }
+
+    @Override
+    protected LLVMValueRef[] getCallArguments(NodeInputList<ValueNode> arguments, CallingConvention.Type callType, ResolvedJavaMethod targetMethod) {
+        LLVMValueRef[] args = super.getCallArguments(arguments, callType, targetMethod);
+        return gen.getCallArguments(args, callType, targetMethod);
+    }
+
+    protected LLVMValueRef emitCondition(LogicNode condition) {
+        if (condition instanceof SafepointCheckNode) {
+            LLVMValueRef threadData = getGenerator().getSpecialRegister(LLVMFeature.THREAD_POINTER_INDEX);
+            threadData = gen.getBuilder().buildIntToPtr(threadData, gen.getBuilder().rawPointerType());
+            LLVMValueRef safepointCounterAddr = gen.getBuilder().buildGEP(threadData, gen.getBuilder().constantInt(Math.toIntExact(Safepoint.getThreadLocalSafepointRequestedOffset())));
+            LLVMValueRef safepointCount = gen.getBuilder().buildAtomicSub(safepointCounterAddr, gen.getBuilder().constantInt(1));
```

Comment:
Since this applies (weak) atomicity guarantees to safepoint checks, can this have any drawbacks in the generated code? I suppose on x86 this could still be emitted as `dec` operation.

Reply:
This is indeed producing suboptimal code (the code generator is not smart enough to emit a `dec` instruction here. That should be a simple call to `buildSub` instead.

Reply:
This can be fixed in the subsequent PRs.

Reply:
@loicottet can you handle this?

Reply:
Yup, I’ll do it when I’m back